feat: peerDAS - implement validator custody

dguenther opened the pull request:
**Motivation**

@hughy and I have a basic implementation of [validator custody](https://github.com/ethereum/consensus-specs/blob/dev/specs/fulu/validator.md) for peerDAS.

We're planning to do more testing on this, but would appreciate a review on the architecture since we're still pretty new!

This relates to #7632 - If it merges, we'll update our PR to account for it.

**Description**

* Centralizes custody values like `sampledGroups` and `custodyGroups` into `CustodyConfig`. `CustodyConfig` is now treated as a singleton.
* Creates a new `advertisedGroupCount` in `CustodyConfig`, used for the custody group count in the node's metadata/ENR.
* Adds `setSamplingGroupCount` and `setAdvertisedGroupCount` to NetworkCore API. Updated by an `EventEmitter` on `CustodyConfig`.
* Adds LocalValidatorRegistry to track connected validators.
* Updates custody requirement in `chain.onForkChoiceFinalized`.

**Not Included**

I'll open separate issues for these if we're okay merging this PR without them.

* Backfilling groups when the target custody group count increases
* Handling changes in other peers' custody group counts
* Race conditions around group count changing during syncing

<!-- A clear and concise general description of the changes of this PR commits -->

<!-- If applicable, add screenshots to help explain your solution -->

<!-- Link to issues: Resolves #111, Resolves #222 -->

Closes #7619

dguenther committed:
Add VALIDATOR_CUSTODY_REQUIREMENT and BALANCE_PER_ADDITIONAL_CUSTODY_GROUP config

hughy committed:
add missing config fields to getSpecCriticalParams

- VALIDATOR_CUSTODY_REQUIREMENT
- BALANCE_PER_ADDITIONAL_CUSTODY_GROUP

hughy committed:
convert CustodyConfig from type to class

- define getCustodyColumnsMeta as private method

dguenther committed:
Add targetCustodyGroupCount and advertisedCustodyGroupCount

hughy committed:
Merge pull request #2 from dguenther/vc-add-cgcs

Add targetCustodyGroupCount and advertisedCustodyGroupCount

dguenther committed:
Fetch sampleSubnets and samplingGroups from CustodyConfig

hughy committed:
compute validators custody requirement

use validator effective balances instead of raw balances assuming spec changes

for now do not use cached state to access effective balances. update if cached
state available when triggering updates to CustodyConfig

return CUSTODY_REQUIREMENT if validatorIndices is empty

dguenther committed:
Use CustodyConfig in PeerDiscovery

hughy committed:
add CustodyConfig to init for NetworkCore in networkCoreWorker

hughy committed:
update CustodyConfig on forkChoiceFinalized events

- set target CGC based on latest finalized state
- convert getValidatorsCustodyRequirement to use CachedBeaconStateAllForks

- TODO: the chain does not currently have a means of identifying validator indices for the node

dguenther committed:
Pass sampling groups into network worker as well

dguenther committed:
Update metadata and ENR with advertised group count

hughy committed:
track local validators

define LocalValidatorRegistry to track indices of validators that have
registered with the node.

read indices of local validators from LocalValidatorRegistry when updating
validator custody requirements.

borrow logic from ValidatorMonitor for registering and pruning validators from
the map in LocalValidatorRegistry.

register validators and prune validators in tandem from LocalValidatorRegistry and ValidatorMonitor.
TODO: re-use validator registry in ValidatorMonitor.

we need to be able to determine the indices of connected validators even if
metrics are disabled.

dguenther committed:
Replace sampledColumns with getSampledColumns

dguenther committed:
Remove custodyGroups from custodyConfig

dguenther committed:
Clean up uses of data column groups

dguenther committed:
Fix from code review

dguenther committed:
Fix dataColumn tests

dguenther committed:
Fix skipped test

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ChainSafe/lodestar?pullRequest=7640) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/ChainSafe/lodestar?pullRequest=7640) before we can accept your contribution.<br/>**0** out of **2** committers have signed the CLA.<br/><br/>:x: dguenther<br/>:x: hughy<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ChainSafe/lodestar?pullRequest=7640) it.</sub>

dguenther committed:
Remove unused CustodyConfig from BaseNetworkInit

dguenther committed:
Simplify diff

nflaig commented on packages/beacon-node/src/api/impl/validator/index.ts:
```diff
@@ -1420,6 +1420,11 @@ export function getValidatorApi(
 
+      // Register validators in LocalValidatorRegistry
+      for (const subscription of subscriptions) {
+        chain.localValidatorRegistry.registerLocalValidator(subscription.validatorIndex);
```
This is not something that needs to be changed for the initial implementation but it seems like with peerdas it would be good to standardize the approach how validators are registered with the beacon node, and deduplicate the caches, re-purposing `prepareBeaconCommitteeSubnet` is not ideal imo.

I think best would be to go with the solution described in https://github.com/ethereum/beacon-APIs/issues/435 which already today would improve the status quo but with validator custody even more so.

eg. we could consolidate these two caches we already have
https://github.com/ChainSafe/lodestar/blob/ae2b1059311c80d12034791be5d5c4872ad68e0e/packages/beacon-node/src/chain/beaconProposerCache.ts#L10

https://github.com/ChainSafe/lodestar/blob/ae2b1059311c80d12034791be5d5c4872ad68e0e/packages/beacon-node/src/execution/builder/cache.ts#L12

I would expect this to be similar for other clients, but first step would be to push on this on the spec and get feedback there.

hughy replied:
Thanks for pointing this out! I'll take a look at using the `BeaconProposerCache` or `ValidatorRegistrationCache` for validator custody rather than introducing a third cache.

hughy replied:
I pushed an update to use `BeaconProposerCache` for now; other client implementations of validator custody (i.e., Prysm) seem to be using `prepareBeaconProposer` to track validator indices for validator custody.

dguenther replied:
Resolving for now since we're not planning to change it for the initial implementation and are pushing on the spec.

dguenther committed:
Update comment

dguenther committed:
Fix Network

dguenther committed:
Simplify diff

dguenther committed:
Simplify discover

dguenther committed:
Remove exports

hughy committed:
remove LocalValidatorRegistry

use existing BeaconProposerCache to access the list of indices of connected
validators

add method to BeaconProposerCache to access validator indices

connected validator clients call 'prepareBeaconProposer' every epoch which adds
or updates an entry in the BeaconProposerCache. we can use this cache instead
of introducing a duplicative data structure

dguenther committed:
Add tests for updateCustodyRequirement

dguenther committed:
Remove unused getSampledGroups

matthewkeil commented on packages/beacon-node/src/chain/chain.ts:
```diff
@@ -1134,11 +1138,12 @@ export class BeaconChain implements IBeaconChain {
 
     const metrics = this.metrics;
-    if (metrics && (slot + 1) % SLOTS_PER_EPOCH === 0) {
+    if ((slot + 1) % SLOTS_PER_EPOCH === 0) {
```
I tend to think this should get reverted to not run the `sleep` if the `metrics` is not present but not a big deal

hughy replied:
Good catch, this is leftover from an earlier change that tracked validator indices in parallel with `ValidatorMonitor`. I'll revert!

matthewkeil commented on packages/beacon-node/src/chain/chain.ts:
```diff
@@ -1188,6 +1193,9 @@ export class BeaconChain implements IBeaconChain {
 
     if (headState) {
       this.opPool.pruneAll(headBlock, headState);
+      // Update custody requirement based on finalized state
+      const validatorIndices = this.beaconProposerCache.getValidatorIndices();
+      this.custodyConfig.updateCustodyRequirement(headState, validatorIndices);
```
This feels like the right idea but wanted to highlight for discussion.  Just thinking out loud here but during periods of non-finality this may pose a challenge.  I suppose if we are going to take the same tack as Manu and just let custody self-prune through expiry then balances that slowly bleed out are not really an issue.... Curious what @nflaig thinks

nflaig replied:
shouldn't we update custody requirements based on effective balance instead of indices? also curious if it's part of spec to do it based on finalized state as this definitely has it's flaws during non-finality

hughy replied:
There were a couple of changes to the spec last week that touched on this: one to specify that we should use the latest finalize state (https://github.com/ethereum/consensus-specs/pull/4211) and a second to specify that we should use the effective balances (https://github.com/ethereum/consensus-specs/pull/4218).

> shouldn't we update custody requirements based on effective balance instead of indices?

I implemented `getValidatorsCustodyRequirement` to calculate effective balance for each validator by its index, let me know if there's a different or better way to do this!

nflaig replied:
Thanks @hughy, maybe I misunderstood what Matt was asking, both the spec PRs you shared make sense to me, non-finality is not something we need to consider for custody requirements

matthewkeil commented on packages/beacon-node/src/chain/seenCache/seenGossipBlockInput.ts:
```diff
@@ -247,15 +247,13 @@ export class SeenGossipBlockInput {
-            true
-          );
+          dataColumnsCache.size >= sampledColumns.length &&
+          sampledColumns.reduce((acc, columnIndex) => acc && dataColumnsCache.has(columnIndex), true);
```
Looks like a lint configuration issue.  Curious why this popped up.  I noticed a few things like this.  Not a huge issue but curious from a repo configuration standpoint for other external contributors.  Do you use vscode?

dguenther replied:
Might just be because the line got shorter? -- I have the biome extension installed (Shows v1.9.4), and I think this got populated from the repository's `.vscode/settings.json`:

```
  "editor.formatOnSave": true,
  "editor.codeActionsOnSave": {
    "source.fixAll.biome": "explicit",
  },
```

matthewkeil commented on packages/beacon-node/src/network/core/networkCore.ts:
```diff
@@ -64,6 +65,8 @@ export type BaseNetworkInit = {
   activeValidatorCount: number;
   initialStatus: phase0.Status;
+  initialSamplingGroupCount: number;
+  initialAdvertisedGroupCount: number;
```
I think we should pass around the `custodyConfig` instead of values it contains and then let the consumer choose what to use.

matthewkeil commented on packages/beacon-node/src/network/metadata.ts:
```diff
@@ -26,6 +26,7 @@ export type MetadataOpts = {
 export type MetadataModules = {
   config: BeaconConfig;
   onSetValue: (key: string, value: Uint8Array) => void;
+  initialAdvertisedGroupCount: number;
```
I think we should be passing the `custodyConfig` into here and then get the value from that object on requests so the `cgc` announcement is up-to-date with our current custody

matthewkeil commented on packages/beacon-node/src/network/peers/discover.ts:
```diff
@@ -46,6 +46,7 @@ export type PeerDiscoveryModules = {
   metrics: NetworkCoreMetrics | null;
   logger: LoggerNode;
   config: BeaconConfig;
+  initialSamplingGroupCount: number;
```
How do you feel about just passing the `custodyConfig` instead of the raw value here.  I didnt notice us instantiating one on the networks side but still reviewing.  Thinking that we may build one on the main thread and one on the network thread and then use the event bus to keep the two in sync.  Not 100% sure that is the right approach and I am open to suggestions if you have another idea.

matthewkeil commented on packages/beacon-node/src/network/peers/discover.ts:
```diff
@@ -109,7 +110,7 @@ export class PeerDiscovery {
   // TODO-das: remove nodeId and sampleSubnets once we remove onlyConnect* flag
   private nodeId: NodeId;
-  private sampleSubnets: number[];
+  private sampleSubnets: CustodyIndex[];
```
Thinking we not use this.  We might want to delete this prop and directly use the custodyConfig, or maybe turn it into a `getter` that pulls from the custodyConfig object.  That way all of the consumers will pull the value from the `custodyConfig` directly and we only have to maintain/update the value on that one object.

matthewkeil commented on packages/beacon-node/src/network/peers/discover.ts:
```diff
@@ -345,6 +338,10 @@ export class PeerDiscovery {
+  setSamplingGroupCount(count: number): void {
+    this.sampleSubnets = getCustodyGroups(this.nodeId, count);
+  }
+
```
As long as we keep the single object sync'd across the worker boundary, and then pull the value from that centralized `custodyConfig` we can get rid of all of the setters.  What do you think?

matthewkeil commented on packages/beacon-node/src/network/peers/discover.ts:
```diff
@@ -431,14 +428,12 @@ export class PeerDiscovery {
-      custodyGroupCount
-    );
-    this.logger.debug("Discovered peer via discv5", {peer: prettyPrintPeerId(peerId), status, custodySubnetCount: custodyGroupCount});
+    const status = this.handleDiscoveredPeer(peerId, multiaddrTCP, attnets, syncnets, custodyGroupCount);
+    this.logger.debug("Discovered peer via discv5", {
+      peer: prettyPrintPeerId(peerId),
+      status,
+      custodySubnetCount: custodyGroupCount,
+    });
```
So weird how all these lint thingyies.... we just recently switched to Biome (which is wonderful btw) but there were some learning curve issues with it.   Would love to get together with you at some point to figure out what settings we need to memorialize or potentially document on or contributor docs!!!  @nazarhussain might have some insight on this

matthewkeil commented on packages/beacon-node/src/network/peers/peerManager.ts:
```diff
@@ -173,17 +173,7 @@ export class PeerManager {
-      this.nodeId,
-      Math.max(this.config.CUSTODY_REQUIREMENT, this.config.NODE_CUSTODY_REQUIREMENT, this.config.SAMPLES_PER_SLOT)
-    );
+    this.sampleSubnets = getCustodyGroups(this.nodeId, modules.initialSamplingGroupCount);
```
Same as `discovery.ts`, I think we should pass in the `custodyConfig` and pull values directly from that centralized object.

matthewkeil commented on packages/beacon-node/src/network/peers/peerManager.ts:
```diff
@@ -173,17 +173,7 @@ export class PeerManager {
-      this.nodeId,
-      Math.max(this.config.CUSTODY_REQUIREMENT, this.config.NODE_CUSTODY_REQUIREMENT, this.config.SAMPLES_PER_SLOT)
-    );
-    // TODO(das): get from custodyConfig or a centralized place every time, instead of computing once here
-    this.samplingGroups = getCustodyGroups(
-      this.nodeId,
-      Math.max(this.config.CUSTODY_REQUIREMENT, this.config.NODE_CUSTODY_REQUIREMENT, this.config.SAMPLES_PER_SLOT)
-    );
```
Not sure we should be deleting this.  We can circle back with the spec and double check.

dguenther replied:
Yeah, I'll roll this back. I'm not sure the original numbers used are 100% correct, but mine definitely aren't, and it's not a related change to validator custody.

dguenther replied:
Rolled back changes here, going to resolve the conversation

matthewkeil commented on packages/beacon-node/src/network/peers/peerManager.ts:
```diff
@@ -565,8 +563,8 @@ export class PeerManager {
-      // ignore samplingGroups for pre-fulu forks
-      forkSeq >= ForkSeq.fulu ? this.samplingGroups : undefined,
+      // ignore sampleSubnets for pre-fulu forks
+      forkSeq >= ForkSeq.fulu ? this.sampleSubnets : undefined,
```
I'm gonna need to dig into this a bit more with the specs.  Lets circle back next week and we can put our heads together on subnets vs groups... Peers will be announcing their groups but need to dig into the code a bit more to see exactly how things are wired up.

dguenther replied:
Going to resolve this for now since I've rolled back the changes, but I made an issue here https://github.com/ChainSafe/lodestar/issues/7658 and set a reminder to follow up on it next week.

matthewkeil commented on packages/beacon-node/src/util/dataColumns.ts:
```diff
@@ -1,41 +1,144 @@
+  }
+
+  getSampledGroupCount(): number {
+    // TODO: Porting this over to match current behavior, but I think this incorrectly mixes units:
+    // SAMPLES_PER_SLOT is in columns, and CUSTODY_GROUP_COUNT is in groups
```
Interesting catch!!  Ill make a note to check into this more deeply

dguenther replied:
It's a bit tough to keep the units straight, but since everything is 1 to 1 in the spec right now, it's probably not super pressing. Here's a couple notes as far as my current understanding:
* Network subscribes to subnets. Subnets are `Column ID % number of subnets`. Groups aren't used.
* Custody happens based on groups. Groups include 1 or more columns

dguenther replied:
I made an issue for generally checking the units here: https://github.com/ChainSafe/lodestar/issues/7658

matthewkeil commented on packages/beacon-node/src/util/dataColumns.ts:
```diff
@@ -1,41 +1,144 @@
+   * Data columns sampled by the node as part of custody sampling
+   * https://github.com/ethereum/consensus-specs/blob/dev/specs/fulu/das-core.md#custody-sampling
+   *
+   * TODO: Consider race conditions if this updates during sync/backfill
```
This is going to be one of the trickier parts of backfill because the SyncChain may not have the same value throughout its life.  We may need to do a double-pass or some other more robust sync methodology.

dguenther replied:
Yeah... I'd lean toward finding a way to attach the number of columns needed to either the sync state or the individual block download's state (probably would then need another pass if the column count changes though).

matthewkeil commented on packages/beacon-node/src/util/dataColumns.ts:
```diff
@@ -1,41 +1,144 @@
+   * TODO: Consider race conditions if this updates during sync/backfill
+   */
+  getSampledColumns(): ColumnIndex[] {
+    return getDataColumns(this.nodeId, this.getSampledGroupCount());
```
My gut tells me that we should cache the values for all of these getters and then update them all a single time.

So maybe instead of a getter we just have the `sampledColumns` value and then in `updateCustodyRequirement` we just calculate all the values a single time.

dguenther replied:
Sounds good, pushed a commit with an update to this!

matthewkeil commented on packages/beacon-node/src/util/dataColumns.ts:
```diff
@@ -1,41 +1,144 @@
+    const newSampledGroupCount = this.getSampledGroupCount();
+
+    if (oldSampledGroupCount !== newSampledGroupCount) {
+      this.emitter.emit(CustodyEvent.samplingGroupCountUpdated, newSampledGroupCount);
```
Now that Ive gotten to the last file I see why you used the emitter from the custodyConfig but I still tend to think we should use one of the existing event busses.  I'm wondering if we can remove `custodyConfig` from the `chain` object and isolate it solely to the `network` and then have all consumers reference it from there.  I see a few places that will not be great though.  Need to think about this a bit.   Trying to get the network object passed into the constructor will create an initialization issue in `packages/beacon-node/src/node/nodejs.ts`.    Hmmmm 🤔

matthewkeil reviewed (changes requested):
Great start !!!!

matthewkeil commented on packages/beacon-node/src/metrics/validatorMonitor.ts:
```diff
@@ -789,6 +790,13 @@ export function createValidatorMonitor(
+    /**
+     * Get the indices of the validators that are being monitored.
+     */
+    getLocalValidatorIndices() {
```
I didnt notice this used anywhere?

hughy replied:
Looks like I added this change accidentally from an earlier approach to accessing validator indices 😅 

Removing!

matthewkeil reviewed (commented):
I spent some time thinking about what you have here and the architecture of Lodestar.  There are a couple things that I think will help long-term.  I also spent some time speaking with @twoeths  to make sure we go down the right path.

Here are some design considerations that we may want to incorporate

- Always using the `custodyConfig` directly to pull values.
- Keeping a `custodyConfig` on the main thread and one on the network thread
- Maintain update of attached validators and balances happening on the chain in `onFinalizedEpoch`
- Pass update of custody to the network (on main thread) via the `ChainEventEmitter`
- Pass update of custody to the network (on network thread) via worker api

Using the method that Tuyen employed we can use the custodyConfig from the chain object directly on the main thread network object which will keep those two in sync.

https://github.com/ChainSafe/lodestar/blob/30505868696b687f726cdf9cbe890b004c9c141f/packages/beacon-node/src/network/network.ts#L123

I'm also thinking about breaking `updateCustodyRequirement` into two parts.  First will be calculating the custody requirement and the second will be updating the `custodyConfig`(s).  This way we can calculate the update in `onFinalizedEpoch` and then event the `custodyRequirement` value to the network object on both main and network threads easily.

```ts
const custodyRequirement = calculateCustodyRequirement(state, validatorIndices, config);
this.custodyConfig.updateTargetCustody(custodyRequirement);
this.events.emit(ChainEvent.UpdateTargetCustody, custodyRequirement);
```

In the network object there would be an `onUpdateTargetCustody` that would be the handler for that event (kind of what you have now but will listen to the chain event bus instead of the custodyConfig object).  The handler would call `this.getApi().updateTargetCustody(custodyRequirement)` to send the value to the network thread so that object could be updated for the PeerManager and other consumers on the network thread.

hughy committed:
reverts changes adjacent to ValidatorMonitor

reverts changes leftover from an earlier approach to tracking local validators
alongside the ValidatorMonitor

- removes 'getLocalValidatorIndices' method from ValidatorMonitor
- sleeps during onClockSlot only if metrics enabled

dguenther committed:
Revert optional operator

dguenther committed:
Revert peerManager changes to samplingGroups

dguenther committed:
Cache values instead of using getters

dguenther committed:
Use ChainEventEmitter instead of CustodyEvent

hughy committed:
add option to disable validator custody

- define CustodyConfigOpts type to encapsulate options for CustodyConfig
- by default, enable validator custody
- add separate method for calculating target custody group count to prevent
  calculation if validator custody disabled
- add CLI flag, 'chain.noValidatorCustody', to disable validator custody

dguenther committed:
Merge remote-tracking branch 'upstream/peerDAS' into peerDAS-validator-custody

hughy committed:
update CustodyConfig separately from computing custody requirement

dguenther committed:
Use CustodyConfig in NetworkConfig

dguenther committed:
Use updateTargetCustodyGroupCount in NetworkConfig

hughy committed:
do not emit custody events if validator custody disabled

- remove CustodyConfigOpts
- remove calculateTargetCustodyGroupCount
- condition all custody update logic in onForkChoiceFinalized on
  noValidatorCustody flag

dguenther commented on packages/beacon-node/src/network/core/networkCore.ts:
```diff
@@ -350,6 +359,15 @@ export class NetworkCore implements INetworkCore {
+
+  async setAdvertisedGroupCount(count: number): Promise<void> {
+    this.networkConfig.setAdvertisedGroupCount(count);
+    this.metadata.cgc = count;
```
Not sure if there's a better way to handle syncing the metadata's CGC to the advertised group count -- maybe passing the metadata controller into the network config.

matthewkeil replied:
I think this is probably ok.  It gets updated with the config so feels like we can be confident that the value is sync'd correctly

dguenther commented on packages/cli/src/options/beaconNodeOptions/chain.ts:
```diff
@@ -320,4 +322,11 @@ Will double processing times. Use only for debugging purposes.",
+  "chain.noValidatorCustody": {
+    type: "boolean",
+    description: "Disable dynamic updates to validator custody requirements",
+    defaultDescription: "false",
```
Is `defaultDescription` needed? Should we default it to `true`?

matthewkeil replied:
not needed but fine to leave that there

dguenther committed:
Cleanup code

dguenther committed:
Merge remote-tracking branch 'upstream/peerDAS' into peerDAS-validator-custody

matthewkeil commented on packages/beacon-node/src/network/metadata.ts:
```diff
@@ -35,16 +35,15 @@ export type MetadataModules = {
       ...ssz.fulu.Metadata.defaultValue(),
-      // TODO: @matthewkeil check if this needs to be updated for custody groups
-      cgc: Math.max(this.config.CUSTODY_REQUIREMENT, this.config.NODE_CUSTODY_REQUIREMENT),
+      cgc: modules.networkConfig.custodyConfig.advertisedCustodyGroupCount,
```
Looks like `custodyConfig` is private.  Should we use the getter?

```suggestion
      cgc: modules.networkConfig.getCustodyConfig().advertisedCustodyGroupCount,
```

matthewkeil commented on packages/beacon-node/src/network/networkConfig.ts:
```diff
@@ -10,12 +10,12 @@ import {NodeId, computeNodeId} from "./subnets/interface.js";
   private readonly nodeId: NodeId;
   private readonly config: BeaconConfig;
-  private custodyConfig: CustodyConfig;
+  readonly custodyConfig: CustodyConfig;
```
ahhh... was reviewing the files in order and see that you changed this.  I think we should use the getter pattern and leave this `private readonly`
```suggestion
  private readonly custodyConfig: CustodyConfig;
```

dguenther replied:
sounds good! pushed an update to use the getter pattern

matthewkeil commented on packages/beacon-node/src/network/networkConfig.ts:
```diff
@@ -26,17 +26,11 @@ export class NetworkConfig {
-   */
-  getCustodyConfig(): CustodyConfig {
-    return this.custodyConfig;
+  setTargetGroupCount(count: number): void {
+    this.custodyConfig.updateTargetCustodyGroupCount(count);
```
```suggestion
 /**
   * Consumer should never mutate returned CustodyConfig
   */
  getCustodyConfig(): CustodyConfig {
    return this.custodyConfig;
  }

  setTargetGroupCount(count: number): void {
    this.custodyConfig.updateTargetCustodyGroupCount(count);
```

matthewkeil commented on packages/beacon-node/src/network/peers/discover.ts:
```diff
@@ -508,11 +503,12 @@ export class PeerDiscovery {
       const peerCustodyColumns = getDataColumns(nodeId, peerCustodyGroupCount);
 
-      const matchingSubnetsNum = this.sampleSubnets.reduce(
+      const sampleSubnets = this.networkConfig.custodyConfig.sampledSubnets;
```
```suggestion
      const sampleSubnets = this.networkConfig.getCustodyConfig().sampledSubnets;
```

matthewkeil commented on packages/beacon-node/src/network/peers/peerManager.ts:
```diff
@@ -428,11 +424,9 @@ export class PeerManager {
-        0
-      );
-      const hasAllColumns = matchingSubnetsNum === this.sampleSubnets.length;
+      const sampleSubnets = this.networkConfig.custodyConfig.sampledSubnets;
```
```suggestion
      const sampleSubnets = this.networkConfig.getCustodyConfig().sampledSubnets;
```

matthewkeil commented on packages/beacon-node/src/network/peers/peerManager.ts:
```diff
@@ -562,7 +556,7 @@ export class PeerManager {
       this.syncnetsService.getActiveSubnets(),
       // ignore samplingGroups for pre-fulu forks
-      forkSeq >= ForkSeq.fulu ? this.samplingGroups : undefined,
+      forkSeq >= ForkSeq.fulu ? this.networkConfig.custodyConfig.sampleGroups : undefined,
```
```suggestion
      forkSeq >= ForkSeq.fulu ? this.networkConfig.getCustodyConfig().sampleGroups : undefined,
```

matthewkeil reviewed (changes requested):
Just the small nit about using the getter so it consistent with how its done for the chain config.  Looks great otherwise!!

dguenther committed:
Access custodyConfig via getter

matthewkeil reviewed (approved):
LGTM!! 🚀

wemeetagain commented:
:tada: This PR is included in [v1.34.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.34.0) :tada:
